processing/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -418,9 +420,11 @@ private Iterator<Entry<RootPartitionRange, Short2ObjectSortedMap<AtomicUpdateGro
       TreeMap<RootPartitionRange, Short2ObjectSortedMap<AtomicUpdateGroup<T>>> stateMap
   )
   {
-    final RootPartitionRange lowFench = new RootPartitionRange(partitionId, partitionId);
+    // remediate submap `fromKey > toKey` issue when partitionId overflows
+    final short partitionIdLowFence = partitionId < 0 ? Short.MAX_VALUE : partitionId;
```

Comment:
I think just checking the argument in the constructor of `RootPartitionRange` is enough. Since `RootPartitionRange` does not accept `startPartitionId` or `endPartitionId` less than 0, we will not have a case where `partitionId` passed to this method is less than 0.
And even if it is, we should throw an exception rather than silently convert it to max value.

Reply:
Thanks for checking on this! @kfaraz , the [constructor](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java#L1055C19-L1055C19) of `RootPartitionRange` that takes short type of `startPartitionId` was defined as `private` and we are forced to use :
```
static RootPartitionRange of(int startPartitionId, int endPartitionId)
    {
      return new RootPartitionRange((short) startPartitionId, (short) endPartitionId);
    }
```
Since `startPartitionId` is in Integer range, when `startPartitionId` > Short.MAV_VALUE ( 32767 ), the casting from int -> short start producing negative number, for example 
<img width="304" alt="image" src="https://github.com/apache/druid/assets/14854898/26e91ef5-6f1e-4a5f-acbc-d99ea5ceb791">
And the loop repeats when `startPartitionId` continued going because casting from Int to Short losses precision:
<img width="201" alt="image" src="https://github.com/apache/druid/assets/14854898/9f6da102-ad6c-41d5-958c-817962ff6794">
We happen to ran into this short overflow scenario described in https://github.com/apache/druid/issues/15091 and our ingestion task for new data was completely broken because of this, throwing exception would still make the ingestion fail.
Here I'm making `startPartitionId` to be Short.MAX_VALUE so that it won't produce `java.lang.IllegalArgumentException: fromKey > toKey` when we do `stateMap.subMap(lowFence, false, highFence, false)` , this is just an remediation.

I believe a better way to handle this is to allow `startPartitionId` and `endPartitionId` to be integer and avoid the problematic precision-loss casting, I can send another PR to this solution if you can confirm why we had this `short` limit originally and it is appropriate to do so.

Best,
Dun

Reply:
Hi @kfaraz thanks again for spending time on this! 
Bumping again just in case you missed my last message , I`m more than glad to if there`s anything need further clarification , I can also connect on a zoom call if it`s convenient for you!

Best Regards,
Dun

Reply:
@dulu98Kurz - I am not entirely sure but `short` was likely chosen to save on memory that storing these ids will take. 32K partitions in one single interval is too high. Can you describe a bit more as to how your cluster ends up in this situation and why is that a genuine scenario? In my experience, almost every time, an interval touching this high number means that compaction is not configured or ingestion is misconfigured.

Reply:
Hi @abhishekagarwal87 , thanks for checking on this!
You are right our investigation suggesting both late-messages from upstream and compactions falling behind, specifically we found there were random late-messages mixed in the kafka topics, it keep adding tiny segments to finalized trunk and eventually goes beyond `short` range and broke live ingestion tasks of new data, setting rejection period was not ideal because it means we will lose data, and because compaction falling behind we can`t afford to wait for it to catch up , I end up hard deleting the problematic time-trunk and then I realized solely relying on compaction seems inadequate.

Admittedly it is not an ideal use-case for Druid to handle random late messages, but it was a really difficult choice when user had to chose between letting ingestion broke vs deleting problematic time trunk.

Reply:
So instead of capping at `short` max, we can possibly cover the gap by:
- Allowing partitionId goes into `Int` range
- Logging error messages to strongly remind user we need to compact/reduce num of segments.

For users who do not have late messages or compaction issues, this change has no impact to them because they won't store more than `short` max segments anyway, so we don`t break the initial intension of saving on mem.

For users who actually can produce segments beyond `short` max, this will buy them more time to compact/reduc number of segments, which may eventually avoid the difficult situation above.

Reply:
From code quality perspective, `Short.toUnsignedInt` is a precision-loss conversion and we used it in 2 files for 18 times, we can simplify the logic and improve readability if we change to `int`
<img width="606" alt="image" src="https://github.com/apache/druid/assets/14854898/47a36432-d6fb-4f7e-9fdf-1610bf54e39b">

Lastly, when partitionId is out of range, the [logic](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java#L421-L423) we use to handle it right now is simply wrong:
```
final RootPartitionRange lowFench = new RootPartitionRange(partitionId, partitionId);
final RootPartitionRange highFence = new RootPartitionRange(Short.MAX_VALUE, Short.MAX_VALUE);
return stateMap.subMap(lowFench, false, highFence, false).entrySet().iterator();
```
For example when:
<img width="304" alt="image" src="https://github.com/apache/druid/assets/14854898/03d6e2c1-ed5d-4e10-b708-e05699fb12f2">
And  `stateMap.subMap(lowFench, false, highFence, false)` will return all entries instead of empty ...

Reply:
Adding the PR that refactor partitionId from short to int so that we can compare the scope of changes
https://github.com/apache/druid/pull/15116